source/opt/merge_return_pass.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "merge_return_pass.h"
+
+#include "instruction.h"
+#include "ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status MergeReturnPass::Process(ir::IRContext *irContext) {
+  InitializeProcessing(irContext);
+
+  context()->InvalidateAnalysesExceptFor(ir::IRContext::kAnalysisDefUse);
+
+  bool modified = false;
+  for (auto &function : *get_module()) {
+    std::vector<ir::BasicBlock *> returnBlocks = collectReturnBlocks(&function);
+    modified |= mergeReturnBlocks(&function, returnBlocks);
+  }
+
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+std::vector<ir::BasicBlock *> MergeReturnPass::collectReturnBlocks(
+    ir::Function *function) {
+  std::vector<ir::BasicBlock *> returnBlocks;
+  for (auto &block : *function) {
+    ir::Instruction &terminator = *block.tail();
+    if (terminator.opcode() == SpvOpReturn ||
+        terminator.opcode() == SpvOpReturnValue) {
+      returnBlocks.push_back(&block);
+    }
+  }
+
+  return std::move(returnBlocks);
+}
+
+bool MergeReturnPass::mergeReturnBlocks(
+    ir::Function *function, const std::vector<ir::BasicBlock *> &returnBlocks) {
+  if (returnBlocks.size() <= 1) {
+    // No work to do.
+    return false;
+  }
+
+  // Create a label for the new return block
+  std::unique_ptr<ir::Instruction> returnLabel(
+      new ir::Instruction(SpvOpLabel, 0u, TakeNextId(), {}));
+  uint32_t returnId = returnLabel->result_id();
+
+  // Create the new basic block
+  std::unique_ptr<ir::BasicBlock> returnBlockUPtr(
+      new ir::BasicBlock(std::move(returnLabel)));
+  function->AddBasicBlock(std::move(returnBlockUPtr));
+  ir::BasicBlock &returnBlock = *(--function->end());
```

Comment:
Seems a bit baroque to pack so much into a single line...  But ok.

Reply:
I'll change this and similar places to use iterator instead. Hopefully, the overall readability increases.